src/Lean/Compiler/JoinPoints.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2022 Henrik Böving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Lean.Compiler.CompilerM
+
+namespace Lean.Compiler
+
+namespace JoinPointChecker
+
+def jpArity (jp : LocalDecl) : Nat :=
+  go jp.value
+where
+  go : Expr → Nat
+  | .lam _ _ body _ => 1 + go body
+  | _ => 0
+
+/--
+Throws an exception if `e` contains a join point.
+-/
+partial def containsNoJp (e : Expr) : CompilerM Unit := do
+  match e.consumeMData with
+  | .proj _ _ struct => containsNoJp struct
+  | .lam .. =>
+    let (_, b) ← visitLambda e
+    containsNoJp b
+  | .letE .. =>
+    let body ← visitLet e (fun e => do containsNoJp e; pure e)
+    containsNoJp body
+  | .app fn arg =>
+    containsNoJp fn
+    containsNoJp arg
+  | .fvar fvarId =>
+    let some decl ← findDecl? fvarId | unreachable!
+    if decl.userName.getPrefix ==`_jp then
+      throwError s!"Join point {decl.userName} in forbidden position"
+  | .sort .. | .forallE .. | .const .. | .lit .. => return ()
+  | .bvar .. | .mvar .. | .mdata .. => unreachable! 
+
+/--
+Obtain all the tail `Expr`s of `e`. Already checking whether non
+tail values contain a join point and throwing an exception if they do.
+-/
+partial def getTails (e : Expr) :  CompilerM (Array Expr) := do
```

Comment:
This function creates a single local context containing all let-declarations and lambda binders `e`. 
This approach does work for the join point checker, but I found it unexpected. I was expecting the functions `getTails` and `checkTail` to be combined into a single `checkJoinPoints` function.